File: frontend/lib/analytics.ts
import { apiRequest } from "./api";

/* =========================================================
   OVERVIEW
========================================================= */

export type AnalyticsOverview = {
  total_students: number;
  eligible_students: number;
  active_drives: number;
  total_applications: number;
  students_placed: number;
  placement_rate: number;
};

export async function getOverview(): Promise<AnalyticsOverview> {
  return apiRequest<AnalyticsOverview>(
    "/analytics/overview"
  );
}

/* =========================================================
   PLACEMENTS
========================================================= */

export type PlacementSummary = {
  students_placed: number;
  placement_rate: number;
  average_package?: number | null;
  highest_package?: number | null;
  lowest_package?: number | null;
};

export async function getPlacements(): Promise<PlacementSummary> {
  return apiRequest<PlacementSummary>(
    "/analytics/placements"
  );
}

/* =========================================================
   APPLICATIONS
========================================================= */

export type ApplicationSummary = {
  total_applications: number;
  selected_applications: number;
  selection_rate: number;
};

export async function getApplications(): Promise<ApplicationSummary> {
  return apiRequest<ApplicationSummary>(
    "/analytics/applications"
  );
}

/* =========================================================
   HIRING TRENDS
========================================================= */

export type HiringTrend = {
  month: string;
  applications: number;
  selections: number;
};

export async function getHiringTrends(): Promise<HiringTrend[]> {
  return apiRequest<HiringTrend[]>(
    "/analytics/hiring-trends"
  );
}

/* =========================================================
   DEPARTMENT-WISE
========================================================= */

export type DepartmentPlacement = {
  department: string;
  total_students: number;
  students_placed: number;
  placement_rate: number;
};

export async function getDepartmentWise(): Promise<
  DepartmentPlacement[]
> {
  return apiRequest<DepartmentPlacement[]>(
    "/analytics/department-wise"
  );
}

/* =========================================================
   YEAR-WISE
========================================================= */

export type YearPlacement = {
  graduation_year: number;
  total_students: number;
  students_placed: number;
  placement_rate: number;
};

export async function getYearWise(): Promise<YearPlacement[]> {
  return apiRequest<YearPlacement[]>(
    "/analytics/year-wise"
  );
}

/* =========================================================
   COMPANY-WISE
========================================================= */

export type CompanyHiring = {
  company_name: string;
  students_placed: number;
  average_package?: number | null;
};

export async function getCompanyWise(): Promise<CompanyHiring[]> {
  return apiRequest<CompanyHiring[]>(
    "/analytics/company-wise"
  );
}

/* =========================================================
   SALARY DISTRIBUTION
========================================================= */

export type SalaryDistribution = {
  salary_range: string;
  students: number;
};

export async function getSalaryDistribution(): Promise<
  SalaryDistribution[]
> {
  return apiRequest<SalaryDistribution[]>(
    "/analytics/salary-distribution"
  );
}

/* =========================================================
   ROLE OFFERS
========================================================= */

export type RoleOffer = {
  role: string;
  offers: number;
};

export async function getRoleOffers(): Promise<RoleOffer[]> {
  return apiRequest<RoleOffer[]>(
    "/analytics/role-offers"
  );
}

/* =========================================================
   COMPANY RECRUITMENT TRENDS
========================================================= */

export type CompanyRecruitmentTrend = {
  month: string;
  company_name: string;
  recruitments: number;
};

export async function getCompanyRecruitmentTrends(): Promise<
  CompanyRecruitmentTrend[]
> {
  return apiRequest<CompanyRecruitmentTrend[]>(
    "/analytics/company-recruitment-trends"
  );
}

/* =========================================================
   HIRING UPDATES
========================================================= */

export type HiringUpdate = {
  company: string;
  selected: number;
};

export async function getHiringUpdates(): Promise<HiringUpdate[]> {
  return apiRequest<HiringUpdate[]>(
    "/analytics/hiring-updates"
  );
}

/* =========================================================
   PLACEMENT INSIGHTS
========================================================= */

export type PlacementInsights = {
  placement_performance: string[];
  attention_required: string[];
  recommendations: string[];
};

export async function getPlacementInsights(): Promise<PlacementInsights> {
  return apiRequest<PlacementInsights>(
    "/analytics/placement-insights"
  );
}

/* =========================================================
   ELIGIBLE STUDENTS
========================================================= */

export type EligibleStudent = {
  student_id: string;
  student_name: string;
  department: string;
  branch?: string | null;
  cgpa?: number | null;
  company?: string | null;
  academic_year?: string | null;
  eligible_drives: number;
  applications: number;
  interview_status?: string | null;
  placement_status?: string | null;
};

/* =========================================================
   ELIGIBLE STUDENT FILTERS
========================================================= */

export interface EligibleStudentFilters {
  department?: string;
  min_cgpa?: number;
  company?: string;
  branch?: string;
  placement_status?: string;
  academic_year?: string;
}

/* =========================================================
   GET ELIGIBLE STUDENTS
========================================================= */

export async function getEligibleStudents(
  filters: EligibleStudentFilters = {}
): Promise<EligibleStudent[]> {
  const params = new URLSearchParams();

  if (filters.department) {
    params.set("department", filters.department);
  }

  if (filters.min_cgpa !== undefined) {
    params.set("min_cgpa", String(filters.min_cgpa));
  }

  if (filters.company) {
    params.set("company", filters.company);
  }

  if (filters.branch) {
    params.set("branch", filters.branch);
  }

  if (filters.placement_status) {
    params.set("placement_status", filters.placement_status);
  }

  if (filters.academic_year) {
    params.set("academic_year", filters.academic_year);
  }

  const queryString = params.toString();

  const endpoint =
    queryString.length > 0
      ? `/analytics/eligible-students?${queryString}`
      : "/analytics/eligible-students";

  return apiRequest<EligibleStudent[]>(endpoint);
}

File: backend/app/schemas/analytics.py
from typing import List, Optional

from uuid import UUID

from pydantic import BaseModel


class AnalyticsOverview(BaseModel):

    total_students: int

    eligible_students: int

    active_drives: int

    total_applications: int

    students_placed: int

    placement_rate: float


class PlacementSummary(BaseModel):

    students_placed: int

    placement_rate: float

    average_package: Optional[float] = None

    highest_package: Optional[float] = None

    lowest_package: Optional[float] = None


class CompanyHiring(BaseModel):

    company_name: str

    students_placed: int

    average_package: Optional[float] = None


class DepartmentPlacement(BaseModel):

    department: str

    total_students: int

    students_placed: int

    placement_rate: float


class YearPlacement(BaseModel):

    graduation_year: int

    total_students: int

    students_placed: int

    placement_rate: float


class ApplicationSummary(BaseModel):

    total_applications: int

    selected_applications: int

    selection_rate: float


class HiringTrend(BaseModel):

    month: str

    applications: int

    selections: int


class EligibleStudent(BaseModel):

    student_id: UUID

    student_name: str

    department: str

    branch: Optional[str] = None

    cgpa: Optional[float] = None

    company: Optional[str] = None

    academic_year: Optional[str] = None

    eligible_drives: int

    applications: int

    interview_status: Optional[str] = None

    placement_status: Optional[str] = None


class SalaryDistribution(BaseModel):

    salary_range: str

    students: int


class RoleOffer(BaseModel):

    role: str

    offers: int


class CompanyRecruitmentTrend(BaseModel):

    month: str

    company_name: str

    recruitments: int


class HiringUpdate(BaseModel):

    company: str

    selected: int


class DriveStatistics(BaseModel):

    drive_id: UUID

    company_name: str

    drive_title: Optional[str] = None

    status: Optional[str] = None

    applications: int

    selected: int

    selection_rate: float

    vacancies: Optional[int] = None


# =========================================================
# PLACEMENT INSIGHTS
# =========================================================

class PlacementInsights(BaseModel):

    placement_performance: List[str]

    attention_required: List[str]

    recommendations: List[str]